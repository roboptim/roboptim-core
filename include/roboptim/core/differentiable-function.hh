// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_DIFFERENTIABLE_FUNCTION_HH
# define ROBOPTIM_CORE_DIFFERENTIABLE_FUNCTION_HH
# include <cstring>
# include <limits>
# include <stdexcept>
# include <utility>

# include <log4cxx/logger.h>

# include <roboptim/core/fwd.hh>

# include <roboptim/core/function.hh>
# include <roboptim/core/portability.hh>

# define ROBOPTIM_DIFFERENTIABLE_FUNCTION_FWD_TYPEDEFS(PARENT)	\
  ROBOPTIM_FUNCTION_FWD_TYPEDEFS (PARENT);			\
  ROBOPTIM_GENERATE_FWD_REFS (gradient);			\
  ROBOPTIM_GENERATE_FWD_REFS (jacobian)

# define ROBOPTIM_DIFFERENTIABLE_FUNCTION_FWD_TYPEDEFS_(PARENT)	\
  ROBOPTIM_FUNCTION_FWD_TYPEDEFS_ (PARENT);			\
  ROBOPTIM_GENERATE_FWD_REFS_ (gradient);			\
  ROBOPTIM_GENERATE_FWD_REFS_ (jacobian)

namespace roboptim
{
  /// \addtogroup roboptim_meta_function
  /// @{

  /// \brief Define an abstract derivable function (\f$C^1\f$).
  ///
  /// A derivable function which provides a way to compute its
  /// gradient/jacobian.
  ///
  /// \f[ f : x \rightarrow f(x) \f]
  /// \f$x \in \mathbb{R}^n\f$, \f$f(x) \in \mathbb{R}^m\f$ where
  /// \f$n\f$ is the input size and \f$m\f$ is the output size.
  ///
  /// Gradient computation is done through the #impl_gradient method
  /// that has to implemented by the concrete class inheriting this
  /// class.
  ///
  /// Jacobian computation is automatically done by concatenating
  /// gradients together, however this naive implementation can be
  /// overridden by the concrete class.
  ///
  ///
  /// The gradient of a \f$\mathbb{R}^n \rightarrow \mathbb{R}^m\f$
  /// function where \f$n > 1\f$ and \f$m > 1\f$ is a matrix.
  /// As this representation is costly, RobOptim considers
  /// these functions as \f$m\f$ \f$\mathbb{R}^n \rightarrow \mathbb{R}\f$
  /// functions. Through that mechanism, gradients are always vectors
  /// and jacobian are always matrices.
  /// When the gradient or the jacobian has to be computed, one has to
  /// precise which of the \f$m\f$ functions should be considered.
  ///
  /// If \f$m = 1\f$, then the function id must always be 0 and can be safely
  /// ignored in the gradient/jacobian computation.
  /// The class provides a default value for the function id so that
  /// these functions do not have to explicitly set the function id.
  template <typename T>
  class GenericDifferentiableFunction : public GenericFunction<T>
  {
  public:
    ROBOPTIM_FUNCTION_FWD_TYPEDEFS_ (GenericFunction<T>);

    /// \brief Gradient type.
    ROBOPTIM_GENERATE_TRAITS_REFS_(gradient);
    /// \brief Jacobian type.
    ROBOPTIM_GENERATE_TRAITS_REFS_(jacobian);

    /// \brief Jacobian size type (pair of values).
    typedef std::pair<size_type, size_type> jacobianSize_t;


    /// \brief Return the gradient size.
    ///
    /// Gradient size is equals to the input size.
    size_type gradientSize () const
    {
      return this->inputSize ();
    }

    /// \brief Return the jacobian size as a pair.
    ///
    /// Gradient size is equals to (output size, input size).
    jacobianSize_t jacobianSize () const
    {
      return std::make_pair (this->outputSize (), this->inputSize ());
    }

    /// \brief Check if the gradient is valid (check size).
    /// \param gradient checked gradient
    /// \return true if valid, false if not
    bool isValidGradient (const_gradient_ref gradient) const
    {
      return gradient.size () == gradientSize ();
    }

    /// \brief Check if the jacobian is valid (check sizes).
    ///
    /// \param jacobian checked jacobian
    /// \return true if valid, false if not
    bool isValidJacobian (const_jacobian_ref jacobian) const
    {
      return jacobian.rows () == jacobianSize ().first
	&& jacobian.cols () == jacobianSize ().second;
    }

    /// \brief Computes the jacobian.
    ///
    /// \param argument point at which the jacobian will be computed
    /// \return jacobian matrix
    jacobian_t jacobian (const_argument_ref argument) const
    {
      jacobian_t jacobian (jacobianSize ().first, jacobianSize ().second);
      jacobian.setZero ();
      this->jacobian (jacobian, argument);
      return jacobian;
    }

    /// \brief Computes the jacobian.
    ///
    /// Program will abort if the jacobian size is wrong before
    /// or after the jacobian computation.
    /// \param jacobian jacobian will be stored in this argument
    /// \param argument point at which the jacobian will be computed
    void jacobian (jacobian_ref jacobian, const_argument_ref argument)
      const
    {
      LOG4CXX_TRACE (this->logger,
		     "Evaluating jacobian at point: " << argument);
      assert (argument.size () == this->inputSize ());
      assert (isValidJacobian (jacobian));

#ifndef ROBOPTIM_DO_NOT_CHECK_ALLOCATION
      bool cur_malloc_allowed = is_malloc_allowed ();
      set_is_malloc_allowed (false);
#endif //! ROBOPTIM_DO_NOT_CHECK_ALLOCATION

      this->impl_jacobian (jacobian, argument);

#ifndef ROBOPTIM_DO_NOT_CHECK_ALLOCATION
      set_is_malloc_allowed (cur_malloc_allowed);
#endif //! ROBOPTIM_DO_NOT_CHECK_ALLOCATION

      assert (isValidJacobian (jacobian));
    }

    /// \brief Computes the gradient.
    ///
    /// \param argument point at which the gradient will be computed
    /// \param functionId function id in split representation
    /// \return gradient vector
    gradient_t gradient (const_argument_ref argument,
			 size_type functionId = 0) const
    {
      assert (functionId < this->outputSize ());
      gradient_t gradient (gradientSize ());
      gradient.setZero ();
      this->gradient (gradient, argument, functionId);
      return gradient;
    }

    /// \brief Computes the gradient.
    ///
    /// Program will abort if the gradient size is wrong before
    /// or after the gradient computation.
    /// \param gradient gradient will be stored in this argument
    /// \param argument point at which the gradient will be computed
    /// \param functionId function id in split representation
    /// \return gradient vector
    void gradient (gradient_ref gradient,
		   const_argument_ref argument,
		   size_type functionId = 0) const
    {
      LOG4CXX_TRACE (this->logger,
		     "Evaluating gradient at point: "
		     << argument
		     << " (function id: " << functionId << ")");
      assert (functionId < this->outputSize ());
      assert (argument.size () == this->inputSize ());
      assert (isValidGradient (gradient));

#ifndef ROBOPTIM_DO_NOT_CHECK_ALLOCATION
      bool cur_malloc_allowed = is_malloc_allowed ();
      set_is_malloc_allowed (false);
#endif //! ROBOPTIM_DO_NOT_CHECK_ALLOCATION

      this->impl_gradient (gradient, argument, functionId);

#ifndef ROBOPTIM_DO_NOT_CHECK_ALLOCATION
      set_is_malloc_allowed (cur_malloc_allowed);
#endif //! ROBOPTIM_DO_NOT_CHECK_ALLOCATION

      assert (isValidGradient (gradient));
    }

    /// \brief Display the function on the specified output stream.
    ///
    /// \param o output stream used for display
    /// \return output stream
    virtual std::ostream& print (std::ostream& o) const;

  protected:
    /// \brief Concrete class constructor should call this constructor.
    ///
    /// \param inputSize input size (argument size)
    /// \param outputSize output size (result size)
    /// \param name function's name
    GenericDifferentiableFunction (size_type inputSize,
				   size_type outputSize = 1,
				   std::string name = std::string ());

    /// \brief Jacobian evaluation.
    ///
    /// Computes the jacobian, can be overridden by concrete classes.
    /// The default behavior is to compute the jacobian from the gradient.
    /// \warning Do not call this function directly, call #jacobian instead.
    /// \param jacobian jacobian will be store in this argument
    /// \param arg point where the jacobian will be computed
    virtual void impl_jacobian (jacobian_ref jacobian, const_argument_ref arg)
      const;

    /// \brief Gradient evaluation.
    ///
    /// Compute the gradient, has to be implemented in concrete classes.
    /// The gradient is computed for a specific sub-function which id
    /// is passed through the functionId argument.
    /// \warning Do not call this function directly, call #gradient instead.
    /// \param gradient gradient will be store in this argument
    /// \param argument point where the gradient will be computed
    /// \param functionId evaluated function id in the split representation
    virtual void impl_gradient (gradient_ref gradient,
				const_argument_ref argument,
				size_type functionId = 0)
      const = 0;

  public:
    /// \brief Get the type-checking flag
    virtual unsigned long flag() const
    {
      return flag_;
    }

    /// \brief Flag representing the Roboptim Function type
    static const unsigned long flag_;
  };

  /// @}

} // end of namespace roboptim

# include <roboptim/core/differentiable-function.hxx>
#endif //! ROBOPTIM_CORE_DIFFERENTIABLE_FUNCTION_HH
