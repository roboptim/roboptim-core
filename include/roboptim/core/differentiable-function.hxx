// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_DIFFERENTIABLE_FUNCTION_HXX
# define ROBOPTIM_CORE_DIFFERENTIABLE_FUNCTION_HXX
# include <boost/algorithm/string/replace.hpp>

# include "roboptim/core/indent.hh"
# include "roboptim/core/util.hh"

namespace roboptim
{
  template <typename T>
  GenericDifferentiableFunction<T>::GenericDifferentiableFunction
  (size_type inputSize, size_type outputSize, std::string name)
    : GenericFunction<T> (inputSize, outputSize, name)
  {
  }

  template <>
  inline void
  GenericDifferentiableFunction<EigenMatrixSparse>::impl_jacobian
  (jacobian_ref jacobian, const_argument_ref argument)
    const
  {
    typedef Eigen::Triplet<value_type> triplet_t;
    std::vector<triplet_t> coefficients;

    for (jacobian_t::Index i = 0; i < this->outputSize (); ++i)
      {
        gradient_t grad = gradient (argument, i);
        for (gradient_t::InnerIterator it (grad); it; ++it)
          {
            const jacobian_t::Index
              idx = static_cast<const jacobian_t::Index> (it.index ());
            coefficients.push_back
              (triplet_t (i, idx, it.value ()));
          }
      }

#ifndef ROBOPTIM_DO_NOT_CHECK_ALLOCATION
    bool cur_malloc_allowed = is_malloc_allowed ();
    set_is_malloc_allowed (true);
#endif //! ROBOPTIM_DO_NOT_CHECK_ALLOCATION

    jacobian.setFromTriplets (coefficients.begin (), coefficients.end ());

#ifndef ROBOPTIM_DO_NOT_CHECK_ALLOCATION
    set_is_malloc_allowed (cur_malloc_allowed);
#endif //! ROBOPTIM_DO_NOT_CHECK_ALLOCATION
  }


  template <typename T>
  void
  GenericDifferentiableFunction<T>::impl_jacobian (jacobian_ref jacobian,
						   const_argument_ref argument)
    const
  {
    for (typename jacobian_t::Index i = 0; i < this->outputSize (); ++i)
       gradient (jacobian.row (i), argument, i);
  }

  template <typename T>
  std::ostream&
  GenericDifferentiableFunction<T>::print (std::ostream& o) const
  {
    if (this->getName ().empty ())
      return o << "Differentiable function";

    std::stringstream ss;
    ss << std::endl;
    char fill = o.fill (' ');
    ss << std::setw ((int)indent (o))
       << ""
       << std::setfill (fill);
    std::string name = this->getName ();
    boost::algorithm::replace_all (name, "\n", ss.str ());

    return o << name << " (differentiable function)";
  }

  template <typename T>
  const unsigned long GenericDifferentiableFunction<T>::flag_ = ROBOPTIM_IS_DIFFERENTIABLE|GenericDifferentiableFunction<T>::parent_t::flag_;
} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_DIFFERENTIABLE_FUNCTION
