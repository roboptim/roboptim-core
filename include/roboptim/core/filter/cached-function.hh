// Copyright (C) 2010 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_FILTER_CACHED_FUNCTION_HH
# define ROBOPTIM_CORE_FILTER_CACHED_FUNCTION_HH
# include <roboptim/core/sys.hh>
# include <roboptim/core/debug.hh>

# include <map>

# include <boost/shared_ptr.hpp>

# include <roboptim/core/n-times-derivable-function.hh>

namespace roboptim
{
  /// \addtogroup roboptim_filter
  /// @{

  namespace detail
  {
    struct ltvector
    {
      bool operator()(const Function::vector_t& v1,
		      const Function::vector_t& v2) const
      {
	Function::size_type it1 = 0;
	Function::size_type it2 = 0;
	
	while (it1 != v1.size () && it2 != v2.size ())
	  {
	    if (fabs (v1(it1) - v2(it2)) < Function::epsilon ())
	      ++it1, ++it2;
	    else if (v1(it1) - v2(it2) < - Function::epsilon ())
	      return true;
	    else return false;
	  }
	return false;
      }
    };
  } // end of namespace detail.

  /// \brief Store previous function computation.
  ///
  /// When an expensive function is called several times at the same
  /// point (exactly!), the cached function prevents useless
  /// computation by caching the function result.
  ///
  /// This filter is experimental in this release.
  template <typename T>
  class CachedFunction : public T
  {
  public:
    /// \brief Import value type.
    typedef typename DifferentiableFunction::value_type value_type;
    /// \brief Import size type.
    typedef typename DifferentiableFunction::size_type size_type;
    /// \brief Import vector type.
    typedef typename DifferentiableFunction::vector_t vector_t;
    /// \brief Import result type.
    typedef typename DifferentiableFunction::result_t result_t;
    /// \brief Import argument type.
    typedef typename DifferentiableFunction::argument_t argument_t;
    /// \brief Import gradient type.
    typedef typename DifferentiableFunction::gradient_t gradient_t;
    /// \brief Import hessian type.
    typedef typename TwiceDifferentiableFunction::hessian_t hessian_t;
    /// \brief Import jacobian type.
    typedef typename DifferentiableFunction::jacobian_t jacobian_t;
    /// \brief Import interval type.
    typedef typename DifferentiableFunction::interval_t interval_t;


    typedef std::map<Function::vector_t, Function::vector_t, detail::ltvector>
      functionCache_t;
    typedef std::map<Function::vector_t, jacobian_t, detail::ltvector>
      jacobianCache_t;
    typedef std::map<Function::vector_t, hessian_t, detail::ltvector>
      hessianCache_t;

    explicit CachedFunction (boost::shared_ptr<const T> fct) throw ();
    ~CachedFunction () throw ();

    void reset () throw ();

  protected:
    virtual void impl_compute (result_t& result, const argument_t& argument)
      const throw ();


    virtual void impl_gradient (gradient_t& gradient,
				const argument_t& argument,
				size_type functionId = 0)
      const throw ();

    virtual void impl_jacobian (jacobian_t& jacobian, const argument_t& arg)
      const throw ();

    virtual void impl_hessian (hessian_t& hessian,
    			       const argument_t& argument,
    			       size_type functionId = 0) const throw ();

    virtual void impl_derivative (gradient_t& derivative,
    				  double argument,
    				  size_type order = 1) const throw ();

  protected:
    boost::shared_ptr<const T> function_;
    mutable std::vector<functionCache_t> cache_;
    mutable std::vector<functionCache_t> gradientCache_;
    mutable jacobianCache_t jacobianCache_;
    mutable std::vector<hessianCache_t> hessianCache_;
  };

  /// @}

} // end of namespace roboptim

# include <roboptim/core/filter/cached-function.hxx>
#endif //! ROBOPTIM_CORE_FILTER_CACHED_FUNCTION_HH
