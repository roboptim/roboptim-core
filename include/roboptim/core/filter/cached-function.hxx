// Copyright (C) 2010 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_FILTER_CACHED_FUNCTION_HXX
# define ROBOPTIM_CORE_FILTER_CACHED_FUNCTION_HXX
# include <boost/format.hpp>

# include <roboptim/core/derivative-size.hh>

namespace roboptim
{
  namespace
  {
    template <typename T>
    std::string cachedFunctionName (const T& fct);

    template <typename T>
    std::string cachedFunctionName (const T& fct)
    {
      boost::format fmt ("%1% (cached)");
      fmt % fct.getName ();
      return fmt.str ();
    }

  } // end of anonymous namespace.

  template <typename T>
  CachedFunction<T>::CachedFunction (boost::shared_ptr<const T> fct) throw ()
    : T (fct->inputSize (), fct->outputSize (), cachedFunctionName (*fct)),
      function_ (fct),
      cache_ (derivativeSize<T>::value),
      gradientCache_ (static_cast<std::size_t> (fct->outputSize ())),
      hessianCache_ (static_cast<std::size_t> (fct->outputSize ()))
  {
  }

  template <typename T>
  CachedFunction<T>::~CachedFunction () throw ()
  {
  }

  template <typename T>
  void
  CachedFunction<T>::reset () throw ()
  {
    cache_.clear ();
    gradientCache_.clear ();
    hessianCache_.clear ();
  }


  template <typename T>
  void
  CachedFunction<T>::impl_compute (result_t& result,
				   const argument_t& argument)
    const throw ()
  {
    typename CachedFunction<T>::functionCache_t::
      const_iterator it = cache_[0].find (argument);
    if (it != cache_[0].end ())
      {
        result = it->second;
        return;
      }
    (*function_) (result, argument);
    cache_[0][argument] = result;
  }


  template <>
  inline void
  CachedFunction<Function>::impl_gradient (gradient_t&,
                                           const argument_t&,
                                           size_type)
    const throw ()
  {
    assert (0);
  }

  template <>
  inline void
  CachedFunction<SparseFunction>::impl_gradient (gradient_t&,
                                                 const argument_t&,
                                                 size_type)
    const throw ()
  {
    assert (0);
  }

  template <typename T>
  void
  CachedFunction<T>::impl_gradient (gradient_t& gradient,
                                    const argument_t& argument,
                                    size_type functionId)
    const throw ()
  {
    typename CachedFunction<T>::gradientCache_t::
      const_iterator it = gradientCache_
      [static_cast<std::size_t> (functionId)].find (argument);
    if (it != gradientCache_[static_cast<std::size_t> (functionId)].end ())
      {
        gradient = it->second;
        return;
      }
    function_->gradient (gradient, argument, functionId);
    gradientCache_[static_cast<std::size_t> (functionId)][argument]
      = gradient;
  }


  template <>
  inline void
  CachedFunction<Function>::impl_jacobian
  (jacobian_t&,
   const argument_t&) const throw ()
  {
    assert (0);
  }

  template <>
  inline void
  CachedFunction<SparseFunction>::impl_jacobian
  (jacobian_t&,
   const argument_t&) const throw ()
  {
    assert (0);
  }


  template <typename T>
  void
  CachedFunction<T>::impl_jacobian
  (jacobian_t& jacobian,
   const argument_t& argument) const throw ()
  {
    typename CachedFunction<T>::jacobianCache_t::
      const_iterator it = jacobianCache_.find (argument);
    if (it != jacobianCache_.end ())
      {
        jacobian = it->second;
        return;
      }
    function_->jacobian (jacobian, argument);
    jacobianCache_[argument] = jacobian;
  }


  template <>
  inline void
  CachedFunction<Function>::impl_hessian
  (hessian_t&, const argument_t&, size_type) const throw ()
  {
    assert (0);
  }

  template <>
  inline void
  CachedFunction<SparseFunction>::impl_hessian
  (hessian_t&, const argument_t&, size_type) const throw ()
  {
    assert (0);
  }

  template <>
  inline void
  CachedFunction<DifferentiableFunction>::impl_hessian
  (hessian_t&, const argument_t&, size_type) const throw ()
  {
    assert (0);
  }

  template <>
  inline void
  CachedFunction<DifferentiableSparseFunction>::impl_hessian
  (hessian_t&, const argument_t&, size_type) const throw ()
  {
    assert (0);
  }


  template <typename T>
  void
  CachedFunction<T>::impl_hessian (hessian_t& hessian,
  				   const argument_t& argument,
  				   size_type functionId)
    const throw ()
  {
    //FIXME: bug detected by Clang. To be fixed.
#ifdef ROBOPTIM_CORE_THIS_DOES_NOT_WORK
    typename CachedFunction<T>::functionCache_t::
      const_iterator it = hessianCache_
      [static_cast<std::size_t> (functionId)].find (argument);
    if (it != hessianCache_[functionId].end ())
      {
        hessian = it->second;
        return;
      }
#endif
    function_->hessian (hessian, argument, functionId);
    hessianCache_[static_cast<std::size_t> (functionId)][argument] = hessian;
  }


  template <>
  inline void
  CachedFunction<Function>::impl_derivative
  (gradient_t&, double, size_type) const throw ()
  {
    assert (0);
  }

  template <>
  inline void
  CachedFunction<SparseFunction>::impl_derivative
  (gradient_t&, double, size_type) const throw ()
  {
    assert (0);
  }

  template <>
  inline void
  CachedFunction<DifferentiableFunction>::impl_derivative
  (gradient_t&, double, size_type) const throw ()
  {
    assert (0);
  }

  template <>
  inline void
  CachedFunction<DifferentiableSparseFunction>::impl_derivative
  (gradient_t&, double, size_type) const throw ()
  {
    assert (0);
  }

  template <>
  inline void
  CachedFunction<TwiceDifferentiableFunction>::impl_derivative
  (gradient_t&, double, size_type) const throw ()
  {
    assert (0);
  }

  template <>
  inline void
  CachedFunction<TwiceDifferentiableSparseFunction>::impl_derivative
  (gradient_t&, double, size_type) const throw ()
  {
    assert (0);
  }

  template <typename T>
  void
  CachedFunction<T>::impl_derivative (gradient_t& derivative,
  				      double argument,
  				      size_type order)
    const throw ()
  {
    typename T::vector_t x (1);
    x[0] = argument;
    typename CachedFunction<T>::functionCache_t::
      const_iterator it = cache_[order].find (x);
    if (it != cache_[order].end ())
      {
        derivative = it->second;
        return;
      }
    function_->derivative (derivative, x, order);
    cache_[order][x] = derivative;
  }

} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_FILTER_CACHED_FUNCTION_HXX
