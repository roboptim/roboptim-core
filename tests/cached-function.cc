// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "shared-tests/fixture.hh"

#include <iostream>

#include <roboptim/core/io.hh>
#include <roboptim/core/differentiable-function.hh>
#include <roboptim/core/util.hh>
#include <roboptim/core/filter/cached-function.hh>

using namespace roboptim;

boost::shared_ptr<boost::test_tools::output_test_stream> output;

struct F : public DifferentiableFunction
{
  F () : DifferentiableFunction (2, 1, "2 * x * x + y")
  {}

  void impl_compute (result_t& res, const argument_t& argument) const throw ()
  {
    (*output) << "computation (not cached)" << std::endl;
    res.setZero ();
    res[0] = 2. * argument[0] * argument[0] + argument[1];
  }

  void impl_gradient (gradient_t& grad, const argument_t& argument,
		      size_type) const throw ()
  {
    (*output) << "gradient computation (not cached)" << std::endl;
    grad.setZero ();
    grad[0] = 4. * argument[0];
    grad[1] = 1.;
  }

  void impl_jacobian (jacobian_t& jacobian, const argument_t& argument)
    const throw ()
  {
    (*output) << "jacobian computation (not cached)" << std::endl;
    jacobian.setZero ();
    jacobian(0,0) = 4. * argument[0];
    jacobian(0,1) = 1.;
  }
};

BOOST_FIXTURE_TEST_SUITE (core, TestSuiteConfiguration)

BOOST_AUTO_TEST_CASE (cached_function)
{
  output = retrievePattern ("cached-function");

  boost::shared_ptr<F> f (new F ());

  CachedFunction<DifferentiableFunction> cachedF (f);

  (*output) << cachedF << ":" << std::endl
	    << std::endl;

  Function::vector_t x (2);
  for (double i = 0.; i < 10.; i += 0.5)
  {
    x[0] = i;
    x[1] = 2*i;
    (*output) << cachedF (x) << std::endl;
    (*output) << cachedF (x) << std::endl;
    BOOST_CHECK_EQUAL ((*f) (x)[0], cachedF (x)[0]);

    (*output) << cachedF.gradient (x) << std::endl;
    (*output) << cachedF.gradient (x) << std::endl;

    (*output) << cachedF.jacobian (x) << std::endl;
    (*output) << cachedF.jacobian (x) << std::endl;
  }

  std::cout << output->str () << std::endl;
  BOOST_CHECK (output->match_pattern ());
}

BOOST_AUTO_TEST_SUITE_END ()
