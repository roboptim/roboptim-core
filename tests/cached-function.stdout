2 * x * x + y (cached) (differentiable function):

computation (not cached)
[1](0)
[1](0)
computation (not cached)
gradient computation (not cached)
[2](0,1)
[2](0,1)
jacobian computation (not cached)
[1,2]((0,1))
[1,2]((0,1))
computation (not cached)
[1](1.5)
[1](1.5)
computation (not cached)
gradient computation (not cached)
[2](2,1)
[2](2,1)
jacobian computation (not cached)
[1,2]((2,1))
[1,2]((2,1))
computation (not cached)
[1](4)
[1](4)
computation (not cached)
gradient computation (not cached)
[2](4,1)
[2](4,1)
jacobian computation (not cached)
[1,2]((4,1))
[1,2]((4,1))
computation (not cached)
[1](7.5)
[1](7.5)
computation (not cached)
gradient computation (not cached)
[2](6,1)
[2](6,1)
jacobian computation (not cached)
[1,2]((6,1))
[1,2]((6,1))
computation (not cached)
[1](12)
[1](12)
computation (not cached)
gradient computation (not cached)
[2](8,1)
[2](8,1)
jacobian computation (not cached)
[1,2]((8,1))
[1,2]((8,1))
computation (not cached)
[1](17.5)
[1](17.5)
computation (not cached)
gradient computation (not cached)
[2](10,1)
[2](10,1)
jacobian computation (not cached)
[1,2]((10,1))
[1,2]((10,1))
computation (not cached)
[1](24)
[1](24)
computation (not cached)
gradient computation (not cached)
[2](12,1)
[2](12,1)
jacobian computation (not cached)
[1,2]((12,1))
[1,2]((12,1))
computation (not cached)
[1](31.5)
[1](31.5)
computation (not cached)
gradient computation (not cached)
[2](14,1)
[2](14,1)
jacobian computation (not cached)
[1,2]((14,1))
[1,2]((14,1))
computation (not cached)
[1](40)
[1](40)
computation (not cached)
gradient computation (not cached)
[2](16,1)
[2](16,1)
jacobian computation (not cached)
[1,2]((16,1))
[1,2]((16,1))
computation (not cached)
[1](49.5)
[1](49.5)
computation (not cached)
gradient computation (not cached)
[2](18,1)
[2](18,1)
jacobian computation (not cached)
[1,2]((18,1))
[1,2]((18,1))
computation (not cached)
[1](60)
[1](60)
computation (not cached)
gradient computation (not cached)
[2](20,1)
[2](20,1)
jacobian computation (not cached)
[1,2]((20,1))
[1,2]((20,1))
computation (not cached)
[1](71.5)
[1](71.5)
computation (not cached)
gradient computation (not cached)
[2](22,1)
[2](22,1)
jacobian computation (not cached)
[1,2]((22,1))
[1,2]((22,1))
computation (not cached)
[1](84)
[1](84)
computation (not cached)
gradient computation (not cached)
[2](24,1)
[2](24,1)
jacobian computation (not cached)
[1,2]((24,1))
[1,2]((24,1))
computation (not cached)
[1](97.5)
[1](97.5)
computation (not cached)
gradient computation (not cached)
[2](26,1)
[2](26,1)
jacobian computation (not cached)
[1,2]((26,1))
[1,2]((26,1))
computation (not cached)
[1](112)
[1](112)
computation (not cached)
gradient computation (not cached)
[2](28,1)
[2](28,1)
jacobian computation (not cached)
[1,2]((28,1))
[1,2]((28,1))
computation (not cached)
[1](127.5)
[1](127.5)
computation (not cached)
gradient computation (not cached)
[2](30,1)
[2](30,1)
jacobian computation (not cached)
[1,2]((30,1))
[1,2]((30,1))
computation (not cached)
[1](144)
[1](144)
computation (not cached)
gradient computation (not cached)
[2](32,1)
[2](32,1)
jacobian computation (not cached)
[1,2]((32,1))
[1,2]((32,1))
computation (not cached)
[1](161.5)
[1](161.5)
computation (not cached)
gradient computation (not cached)
[2](34,1)
[2](34,1)
jacobian computation (not cached)
[1,2]((34,1))
[1,2]((34,1))
computation (not cached)
[1](180)
[1](180)
computation (not cached)
gradient computation (not cached)
[2](36,1)
[2](36,1)
jacobian computation (not cached)
[1,2]((36,1))
[1,2]((36,1))
computation (not cached)
[1](199.5)
[1](199.5)
computation (not cached)
gradient computation (not cached)
[2](38,1)
[2](38,1)
jacobian computation (not cached)
[1,2]((38,1))
[1,2]((38,1))
